### DISCUSSION TITLE
How does the ability to remember past messages work with the API?

### DISCUSSION THREAD

**Discussion by super-15** [waylaidwanderer/node-chatgpt-api#194] [Discussions]
Hey, I was wondering how does it work for my own similar project. I tried putting the past messages in the new messages but that uses up a bunch of tokens.

**Comment by waylaidwanderer** [comment]
By default, it reserves up to 3k tokens for the conversation history. Any older messages that would exceed this limit are discarded.

**Reply by super-15** [reply]
Oh okay, thanks. Also how does the "typing" effect where it shows what it has generated so far work?

**Reply by waylaidwanderer** [reply]
Responses are generated by the AI one token at a time, so I'm just displaying it as it comes in. Same way ChatGPT works.

**Reply by super-15** [reply]
I tried it with JavaScript using the fetch function to make the api request, and it just sends the JSON when it finishes, not each token live.

**Reply by waylaidwanderer** [reply]
You need to set `stream` to `true` and read the body as a buffer IIRC.

**Reply by waylaidwanderer** [reply]
Look at my implementation in the demos folder.
